### ISSUE TITLE
<function> not working correctly with open/closed interval domains

### ISSUE THREAD

**Issue by rogness** [Doenet/DoenetTools#1385]
I believe these are known issues, but I figured I'd submit here just to keep track.  I haven't dug into code to see if this is all due to one snippet, but we could split it into multiple issues if need be.

Short version: functions don't always work correctly with specified domains, and the behavior is highly dependent on whether endpoints are open or closed.  For example

````
<function name="f" domain="(0,4pi)">cos(x)</function>
<function name="g" domain="[0,4pi)">cos(x)</function>
<function name="h" domain="[0,4pi]">cos(x)</function>

<graph>$f</graph>
<graph>$g</graph>
<graph>$h</graph>
````

The graph of f runs from 0 to 4pi, but the graphs of g and h ignore the domain, and graph the entire function.  Other issues:

````
<m>f(0) = $$f(0)</m>
````
returns f(0)=1. Mathematically it should be undefined, but maybe this is a useful feature in some contexts?  Minima and Maxima have some issues, too:

````
<p>$(f{prop="maxima"})</p>
````

Returns (3.289359698*10^(-12), 1), (6.283185307,1).  That's incorrect, but it's incorrect in a common way, i.e. lots of numerical routines can get confused around 0 and accidentally report an extremum at a small value.  (Doenet does correctly ignore the local extremum at x=4pi.)  However, both of

````
<p>$(g{prop="maxima"})</p>
<p>$(h{prop="maxima"})</p>
````

revert back to finding extrema on [-100,100].  Half open / half closed intervals are weird and might be more of a future project, but when the domain is a closed interval [a,b], I think users will expect Doenet to focus on that interval, not [-100,100]?  (Especially because it _does_ focus on the interior of an open interval.)

The numerical computation issues are less common with polynomials than trig functions, but the domain issues are still there.  For example,

````
<function name="f" domain="[0,2]">(x+1)(x-2)(x-4)</function>
<p>$(f{prop="minima"})</p>
````

returns the minimum at x=3.11963, which is outside of the domain.  (But within [-100,100].)

**Comment by dqnykamp** [comment]
I just checked the code, and domain currently takes a list of tuples in parens (a,b).  (It is a list, presumably, so one could represent rectangular domains of multivariable functions.) It completely ignores values like "[3,4]", which explains the described behavior.

The code does seem to be treating the tuple as a closed interval.

Fixing the representation to parse intervals would be straightforward.

Correctly determining (local) extrema at the endpoints of the domain, on the other hand, will be tricky.  If there is an extremum at an endpoint, a numerical algorithm is likely to find a value that is slightly off the actual value.  In this case, the extremum will be counted if the numerical error happens to push the value into the domain.

**Comment by rogness** [comment]
Followup note / potential workaround for now:

````
<function name="f" domain="(0,2)">(x+1)(x-2)(x-4)</function>
<p>Max on [0,2]: <max>$$f(0) $(f{prop="maximumValues"}) $$f(2)</max>.</p>
````

That code finds the absolute maximum of f on [0,2], and has the added "feature" that it mirrors the standard Closed Interval Method from calculus: find the local maxima in the interior of the open interval, then check the endpoints, and then choose the largest function value. But it's a bit self-contradictory: you have to use an open domain to get the maxima between 0 and 2, which means f(0) and f(2) shouldn't be defined.

Also, the corresponding code to find the absolute minimum of f on [0,2] fails:

````
<p>Max on [0,2]: <min>$$f(0) $(f{prop="minimumValues"}) $$f(2)</min>.</p>
````

That polynomial has no local minimum on (0,2), so the min tag returns (8, _, 0). I don't know the best resolution to that. Maybe instead of changing what "minimumValues" reports from the function component, min should just ignore the null list if there are other numbers it can choose?

**Comment by dqnykamp** [comment]
The behavior of <min> was a flaw.  Fixed in #1388.
